op-program/client/interop/cnosolidate.go

```diff
@@ -0,0 +1,227 @@
+		panic(fmt.Sprintf("unexpected error: failed to create dependency set: %v", err))
+	}
+	return &consolidateCheckDeps{oracle: oracle, heads: heads, depset: depset}
+}
+
+func (d *consolidateCheckDeps) Check(
+	chain supervisortypes.ChainID,
+	blockNum uint64,
+	timestamp uint64,
+	logIdx uint32,
+	logHash common.Hash,
+) (includedIn supervisortypes.BlockSeal, err error) {
+	head := d.heads[chain]
+	if head == nil {
+		return supervisortypes.BlockSeal{}, fmt.Errorf("head not found for chain %v", chain)
+	}
+	// We can assume the oracle has the block the executing message is in
+	block := BlockByNumber(d.oracle, head, blockNum, chain.ToBig().Uint64())
+	return supervisortypes.BlockSeal{
+		Hash:      block.Hash(),
+		Number:    block.NumberU64(),
+		Timestamp: block.Time(),
+	}, nil
+}
+
+func (d *consolidateCheckDeps) IsCrossUnsafe(chainID supervisortypes.ChainID, block eth.BlockID) error {
+	// TODO: assumed to be cross-unsafe
+	// But if the block a future block, then retunr an error
+	return nil
+}
+
+func (d *consolidateCheckDeps) IsLocalUnsafe(chainID supervisortypes.ChainID, block eth.BlockID) error {
+	// Always assumed to be local-unsafe
+	return nil
+}
+
+func (d *consolidateCheckDeps) ParentBlock(chainID supervisortypes.ChainID, parentOf eth.BlockID) (parent eth.BlockID, err error) {
+	head := d.heads[chainID]
+	if head == nil {
+		return eth.BlockID{}, fmt.Errorf("head not found for chain %v", chainID)
+	}
+	block := BlockByNumber(d.oracle, head, parentOf.Number-1, chainID.ToBig().Uint64())
+	return eth.BlockID{
+		Hash:   block.Hash(),
+		Number: block.NumberU64(),
+	}, nil
+}
+
+func (d *consolidateCheckDeps) OpenBlock(
+	chainID supervisortypes.ChainID,
+	blockNum uint64,
+) (ref eth.BlockRef, logCount uint32, execMsgs map[uint32]*supervisortypes.ExecutingMessage, err error) {
+	head := d.heads[chainID]
+	if head == nil {
+		return eth.BlockRef{}, 0, nil, fmt.Errorf("head not found for chain %v", chainID)
+	}
+	block := BlockByNumber(d.oracle, head, blockNum, chainID.ToBig().Uint64())
+	ref = eth.BlockRef{
+		Hash:   block.Hash(),
+		Number: block.NumberU64(),
+	}
+	receipts := d.oracle.ReceiptsByBlockHash(block.Hash(), chainID.ToBig().Uint64())
+	execs, err := ReceiptsToExecutingMessages(receipts)
+	if err != nil {
+		return eth.BlockRef{}, 0, nil, err
+	}
+	execMsgs = make(map[uint32]*supervisortypes.ExecutingMessage, len(execs))
+	for _, exec := range execs {
+		execMsgs[exec.LogIdx] = exec
+	}
+	return ref, uint32(len(execs)), execMsgs, nil
+}
+
+func (d *consolidateCheckDeps) DependencySet() depset.DependencySet {
+	return d.depset
+}
+
+func BlockByNumber(oracle l2.Oracle, head *ethtypes.Block, blockNum uint64, chainID uint64) *ethtypes.Block {
+	if head.NumberU64() < blockNum {
+		return nil
+	}
+	for {
+		if head.NumberU64() == blockNum {
+			return head
+		}
+		if blockNum == 0 {
+			return nil
+		}
+		// TODO: maintain a cache at the earliest head
+		head = oracle.BlockByHash(head.ParentHash(), chainID)
+	}
+}
```

Comment:
We should make this reuse the code in `OracleBackedL2Chain` which does the caching already:
https://github.com/ethereum-optimism/optimism/blob/ca583f7fbdd7d84b1e124c9e7ca2ac207efcd657/op-program/client/l2/engine_backend.go#L84-L100)

It could probably be used as-is I think and would do the OutputRoot lookup and track the L2 head for you as well.  If not, could probably just extract the block by number functionality and cache to another struct that it can include and you can then use that directly.

Reply:
aye. I've copied that logic, including caching, over to consolidation.